pkg/services/ngalert/state/manager_test.go

```diff
@@ -282,2076 +280,999 @@ func TestDashboardAnnotations(t *testing.T) {
-				Title:        "test_title",
-				UID:          "test_alert_rule_uid_2",
-				NamespaceUID: "test_namespace_uid",
-				Data: []models.AlertQuery{{
-					RefID:         "A",
-					DatasourceUID: "datasource_uid_1",
-				}},
-				Annotations:     map[string]string{"annotation": "test"},
-				Labels:          map[string]string{"label": "test"},
-				IntervalSeconds: 10,
-				For:             1 * time.Minute,
-				ExecErrState:    models.ErrorErrState,
-			},
-			evalResults: []eval.Results{
-				{
-					eval.Result{
-						Instance:           data.Labels{"instance_label": "test"},
-						State:              eval.Normal,
-						EvaluatedAt:        evaluationTime,
-						EvaluationDuration: evaluationDuration,
-					},
+		{
+			desc:      "normal -> error when result is Error and ExecErrState is Error",
+			alertRule: baseRuleWith(models.WithForNTimes(6), models.WithErrorExecAs(models.ErrorErrState)),
+			evalResults: map[time.Time]eval.Results{
+				t1: {
+					newResult(eval.WithState(eval.Normal), eval.WithLabels(labels1)),
 				},
-				{
-					eval.Result{
-						Instance:           data.Labels{"instance_label": "test"},
-						Error:              expr.MakeQueryError("A", "", errors.New("this is an error")),
-						State:              eval.Error,
-						EvaluatedAt:        evaluationTime.Add(10 * time.Second),
-						EvaluationDuration: evaluationDuration,
-					},
+				t2: {
+					newResult(eval.WithError(expr.MakeQueryError("A", "datasource_uid_1", errors.New("this is an error"))), eval.WithLabels(labels1)), // TODO fix it because error labels are different
 				},
 			},
 			expectedAnnotations: 1,
-			expectedStates: map[string]*state.State{
-				`[["__alert_rule_namespace_uid__","test_namespace_uid"],["__alert_rule_uid__","test_alert_rule_uid_2"],["alertname","test_title"],["instance_label","test"],["label","test"]]`: {
-					AlertRuleUID: "test_alert_rule_uid_2",
-					OrgID:        1,
-					CacheID:      `[["__alert_rule_namespace_uid__","test_namespace_uid"],["__alert_rule_uid__","test_alert_rule_uid_2"],["alertname","test_title"],["instance_label","test"],["label","test"]]`,
-					Labels: data.Labels{
-						"__alert_rule_namespace_uid__": "test_namespace_uid",
-						"__alert_rule_uid__":           "test_alert_rule_uid_2",
-						"alertname":                    "test_title",
-						"label":                        "test",
-						"instance_label":               "test",
-						"datasource_uid":               "datasource_uid_1",
-						"ref_id":                       "A",
-					},
-					Values: make(map[string]float64),
-					State:  eval.Error,
-					Error:  expr.MakeQueryError("A", "", errors.New("this is an error")),
+			expectedStates: []*state.State{
+				{
+					CacheID: func() string {
+						lbls := models.InstanceLabels(labels["system + rule + labels1"])
+						r, err := lbls.StringKey()
+						if err != nil {
+							panic(err)
+						}
+						return r
+					}(),
+					Labels: mergeLabels(labels["system + rule"], data.Labels{
```

Comment:
Should be `labels["system + rule + labels1"]` right?

Reply:
no, it's error labels which do not exist. I decided to keep it this way because error is defined with literals

Reply:
ah, you're right! This is not fixed yet but I am thinking about it as it is